Add shellcheck to the justfile validate recipe

The `just validate` command only runs YAML schema validation. Adding `shellcheck` to this recipe would let developers run the same lint checks locally with a single command, consistent with the CI workflow.

_Follow-up from #45_

# Implementation Plan

Now I have comprehensive context. Let me produce the implementation plan.

---

## Implementation Plan: Issue #218 — Add shellcheck to `just validate`

### 1. Objective

Add `shellcheck` to the `just validate` recipe so that running a single command locally runs both YAML schema validation and shell script linting — matching what the CI workflow does. Also add shellcheck to the CI `validate.yml` workflow for parity.

---

### 2. Approach

- **Add `.shellcheckrc` to the repo root.** A config already exists in `.worktrees/issue-45/.shellcheckrc`; it needs to land in the actual repo root so shellcheck auto-discovers it.
- **Extend the `validate` recipe in `justfile`** to call `shellcheck` on all `scripts/` shell files and `hooks/pre-commit` after YAML validation succeeds.
- **Extend `.github/workflows/validate.yml`** to install shellcheck and run the same lint step in CI, completing the local↔CI parity.
- **No new wrapper scripts** — shellcheck is called inline in the justfile recipe (consistent with how the current `validate` recipe works) and directly in the CI step.
- **No test file needed** — shellcheck is a well-understood external tool; there is no test harness to write for the invocation itself. The existing scripts already use `# shellcheck source=` directives and `set -euo pipefail`, so they should pass without changes.

Key decision: run shellcheck in the same `validate` recipe rather than a separate `lint` recipe, because the issue explicitly asks for parity with CI in a single command.

---

### 3. Files to Create

| File | Description |
|------|-------------|
| `.shellcheckrc` | Shellcheck configuration (copied from `.worktrees/issue-45/.shellcheckrc`): `severity=warning`, `shell=bash`, `disable=SC1091` |

---

### 4. Files to Modify

#### `justfile`
- In the `validate` recipe, after the existing YAML validation block, add a shellcheck step that lints all `scripts/**/*.sh` files and `hooks/pre-commit`.
- Add a guard at the top of the recipe (alongside existing `yq`/`check-jsonschema` guards) that checks `shellcheck` is installed and prints a helpful install message if not.

Specific change (after current YAML validation block ends):

```just
# Lint shell scripts
echo "==> Running shellcheck on shell scripts..."
shellcheck scripts/*.sh scripts/lib/*.sh hooks/pre-commit
echo "    shellcheck passed"
```

And add to the dependency check block at the recipe top:

```just
command -v shellcheck >/dev/null 2>&1 || { echo "ERROR: shellcheck not found. Install: apt-get install shellcheck / brew install shellcheck"; exit 1; }
```

#### `.github/workflows/validate.yml`
- Add a step to install `shellcheck` (available via `apt-get` on `ubuntu-latest` runners).
- Add a step that runs `shellcheck scripts/*.sh scripts/lib/*.sh hooks/pre-commit` using the same glob pattern as the justfile recipe.

---

### 5. Implementation Order

1. **Add `.shellcheckrc` to repo root** — copy the config from `.worktrees/issue-45/.shellcheckrc` verbatim.
2. **Verify scripts already pass** — run `shellcheck scripts/*.sh scripts/lib/*.sh hooks/pre-commit` locally to confirm no pre-existing errors before modifying any files.
3. **Modify `justfile`** — extend the `validate` recipe with the shellcheck dependency guard and the shellcheck invocation step.
4. **Modify `.github/workflows/validate.yml`** — add `apt-get install -y shellcheck` in the setup step and add a `Run shellcheck` job step.
5. **Run `just validate` locally** — confirm end-to-end: YAML validation passes, then shellcheck passes.
6. **Commit** — stage `.shellcheckrc`, `justfile`, and `.github/workflows/validate.yml`.

---

### 6. Verification

| Check | Command |
|-------|---------|
| Shellcheck passes on all scripts | `shellcheck scripts/*.sh scripts/lib/*.sh hooks/pre-commit` |
| `just validate` exits 0 end-to-end | `just validate` |
| Validate recipe shows shellcheck output | Inspect stdout for `==> Running shellcheck...` and `shellcheck passed` |
| CI workflow step appears and passes | Push to a PR branch; check the `Validate` workflow run in GitHub Actions |
| Missing shellcheck shows useful error | `PATH= just validate` (or uninstall shellcheck) and confirm the error message prints install instructions |

---

### 7. Skills Used

- **`hephaestus:repo-analyze-quick`** — not invoked directly, but the Explore agent performed equivalent codebase discovery
- **`Explore` subagent** — used to read `justfile`, all `scripts/`, `.github/workflows/validate.yml`, `hooks/pre-commit`, and `.worktrees/issue-45/.shellcheckrc` to establish the current state before planning any changes

---
*Generated by Claude Code Planner*

--- a/.github/workflows/apply.yml
+++ b/.github/workflows/apply.yml
@@ -9,6 +9,9 @@ on:
       - 'pixi.toml'
       - 'pixi.lock'
 
+permissions:
+  contents: read
+
 jobs:
   apply:
     name: Reconcile Desired State
@@ -17,6 +20,12 @@ jobs:
       AGAMEMNON_URL: ${{ secrets.AGAMEMNON_URL }}
       AGAMEMNON_API_KEY: ${{ secrets.AGAMEMNON_API_KEY }}
     steps:
+      - name: Mask secrets
+        run: |
+          if [[ -n "$AGAMEMNON_API_KEY" ]]; then
+            echo "::add-mask::$AGAMEMNON_API_KEY"
+          fi
+
       - uses: actions/checkout@v4
 
       - name: Cache pixi environment

--- a/justfile
+++ b/justfile
@@ -21,6 +21,10 @@ agamemnon_url := env_var_or_default("AGAMEMNON_URL", "http://localhost:8080")
 status HOST=host:
     AGAMEMNON_URL={{agamemnon_url}} bash scripts/status.sh {{HOST}}
 
+# Show desired vs actual state for a named fleet
+status-fleet FLEET:
+    AGAMEMNON_URL={{agamemnon_url}} bash scripts/status.sh --fleet {{FLEET}}
+
 # Display the last reconciliation report (JSON)
 report:
     #!/usr/bin/env bash
@@ -94,6 +98,12 @@ validate:
         exit 1
     fi
     echo "All YAML files valid."
+    echo ""
+    echo "Checking name uniqueness..."
+    pixi run lint-names
+    echo ""
+    echo "Running shellcheck..."
+    pixi run lint-shell
 
 # =============================================================================
 # Testing

--- a/pixi.toml
+++ b/pixi.toml
@@ -12,8 +12,10 @@ jq   = ">=1.6"
 # Install manually: see scripts/lib/reconcile.sh or CI workflows.
 
 [tasks]
-status   = "just status"
-plan     = "just plan"
-apply    = "just apply"
-validate = "just validate"
-export   = "just export"
+status        = "just status"
+plan          = "just plan"
+apply         = "just apply"
+validate      = "just validate"
+export        = "just export"
+lint-names    = "bash scripts/lint-names.sh"
+lint-shell    = "bash scripts/lint-shell.sh"

--- a/scripts/lint-names.sh
+++ b/scripts/lint-names.sh
@@ -1,0 +1,71 @@
+#!/usr/bin/env bash
+# scripts/lint-names.sh — Check for duplicate metadata.name across agents and fleets
+#
+# This script validates that all metadata.name values are unique across all agent
+# and fleet YAML files in the agents/ and fleets/ directories.
+#
+# Usage:
+#   ./scripts/lint-names.sh
+#
+# Exit codes:
+#   0 = all names are unique
+#   1 = duplicate names found
+
+set -euo pipefail
+
+SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+REPO_ROOT="$(cd "${SCRIPT_DIR}/.." && pwd)"
+
+ERRORS=0
+
+if ! command -v yq &>/dev/null; then
+    echo "ERROR: yq is required for name validation." >&2
+    echo "  Install: https://github.com/mikefarah/yq" >&2
+    exit 1
+fi
+
+echo "Checking metadata.name uniqueness across all agents and fleets..."
+
+# Associative array to track names for uniqueness check
+declare -A metadata_names
+
+# Find all YAML files (excluding templates)
+while IFS= read -r -d '' file; do
+    [[ "$file" == *"/_templates/"* ]] && continue
+
+    # Extract metadata.name
+    name="$(yq eval '.metadata.name // ""' "$file" 2>/dev/null || true)"
+
+    # Track names for uniqueness check (only non-empty names to avoid false positives)
+    if [[ -n "$name" ]]; then
+        metadata_names["$name"]+=" $file"
+    fi
+
+done < <(find "${REPO_ROOT}/agents" "${REPO_ROOT}/fleets" \
+    -name "*.yaml" -print0 2>/dev/null)
+
+# Cross-file name uniqueness check
+for name in "${!metadata_names[@]}"; do
+    files_with_name="${metadata_names[$name]}"
+    # Count non-empty tokens (file paths are separated by spaces)
+    count=0
+    for f in $files_with_name; do
+        [[ -n "$f" ]] && count=$((count + 1))
+    done
+    if [[ $count -gt 1 ]]; then
+        echo "ERROR: duplicate metadata.name '${name}' found in:"
+        for f in $files_with_name; do
+            [[ -n "$f" ]] && echo "  - ${f#"${REPO_ROOT}/"}"
+        done
+        ERRORS=$((ERRORS + 1))
+    fi
+done
+
+if [[ $ERRORS -gt 0 ]]; then
+    echo ""
+    echo "lint-names: ${ERRORS} duplicate name(s) found."
+    exit 1
+fi
+
+echo "All metadata.name values are unique."
+exit 0

--- a/scripts/lint-shell.sh
+++ b/scripts/lint-shell.sh
@@ -1,0 +1,42 @@
+#!/usr/bin/env bash
+# scripts/lint-shell.sh — Run shellcheck on all shell scripts
+#
+# This script runs shellcheck (with --severity=warning) on all shell scripts
+# found in the scripts/, tests/, and hooks/ directories.
+#
+# Usage:
+#   ./scripts/lint-shell.sh
+#
+# Exit codes:
+#   0 = no shellcheck warnings or errors
+#   1 = shellcheck warnings or errors found
+
+set -euo pipefail
+
+SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+REPO_ROOT="$(cd "${SCRIPT_DIR}/.." && pwd)"
+
+# Find shellcheck: try system first, then pixi lint environment
+SHELLCHECK_CMD="shellcheck"
+if ! command -v shellcheck &>/dev/null; then
+    if [[ -f "${REPO_ROOT}/.pixi/envs/lint/bin/shellcheck" ]]; then
+        SHELLCHECK_CMD="${REPO_ROOT}/.pixi/envs/lint/bin/shellcheck"
+    else
+        echo "ERROR: shellcheck is required for shell linting." >&2
+        echo "  Install: https://www.shellcheck.net/wiki/Install" >&2
+        echo "  Or run: pixi run lint-shell" >&2
+        exit 1
+    fi
+fi
+
+echo "Running shellcheck on all shell scripts..."
+echo ""
+
+# Find and check all shell scripts
+find "${REPO_ROOT}/scripts" "${REPO_ROOT}/tests" "${REPO_ROOT}/hooks" \
+    -name '*.sh' -o -name 'pre-commit' \
+    | sort \
+    | xargs "$SHELLCHECK_CMD" --severity=warning
+
+echo ""
+echo "Shell script linting complete."

